clippy_lints/src/hash_collision.rs

```diff
@@ -0,0 +1,91 @@
+use clippy_utils::diagnostics::span_lint_and_note;
+use clippy_utils::last_path_segment;
+use clippy_utils::ty::is_type_diagnostic_item;
+use rustc_hir::ExprKind;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::declare_lint_pass;
+use rustc_span::symbol::sym;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// When two items are inserted into a `HashMap` with the same key,
+    /// the second item will overwrite the first item.
+    ///
+    /// ### Why is this bad?
+    /// This can lead to data loss.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// # use std::collections::HashMap;
+    /// let example = HashMap::from([(5, 1), (5, 2)]);
+    /// ```
+    #[clippy::version = "1.79.0"]
+    pub HASH_COLLISION,
+    suspicious,
+    "`HashMap` with two identical keys loses data"
+}
+
+declare_lint_pass!(HashCollision => [HASH_COLLISION]);
+
+impl<'tcx> LateLintPass<'tcx> for HashCollision {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx rustc_hir::Expr<'tcx>) {
+        if has_hash_collision(cx, expr) {
+            span_lint_and_note(
+                cx,
+                HASH_COLLISION,
+                expr.span,
+                "this `HashMap` has a hash collision and will lose data",
+                None,
+                "consider using a different keys for all items",
+            );
+        }
+    }
+}
+
+// TODO: Also check for different sources of hash collisions
+// TODO: Maybe other types of hash maps should be checked as well?
+fn has_hash_collision(cx: &LateContext<'_>, expr: &rustc_hir::Expr<'_>) -> bool {
+    // If the expression is a call to `HashMap::from`, check if the keys are the same
+    if let ExprKind::Call(func, args) = &expr.kind
+        // First check for HashMap::from
+        && let ty = cx.typeck_results().expr_ty(expr)
+        && is_type_diagnostic_item(cx, ty, sym::HashMap)
+        && let ExprKind::Path(func_path) = func.kind
+        && last_path_segment(&func_path).ident.name == sym::from
+        // There should be exactly one argument to HashMap::from
+        && args.len() == 1
+        // which should be an array
+        && let ExprKind::Array(args) = &args[0].kind
+    // Then check the keys
+    {
+        // Put all keys in a vector
+        let mut literals = Vec::new();
```

Comment:
You could use `SpanlessEq` to correctly lint on arbitrary key expressions.

Also, why not return a (possibly empty) set of duplicate key spans instead of a plain bool? That way we could better pinpoint the error.

Reply:
Thanks for point out `SpanlessEq`. The function now returns `Option<Vec<(Expr, Expr)>>`, so all pairs of found duplicates. I should probably someday change it to return all Expressions with their duplicates instead of split over multiple pairs.